### ISSUE TITLE
Converge issue when using DeepSpeed FusedLamb

### ISSUE THREAD

**Issue by LiweiPeng** [microsoft/DeepSpeed#182]
Hi, I tried to use DeepSpeed's FusedLamb to run BERT pretrain, but I couldn't get a good converge, especially when compared with using [Apex's FusedLamb](https://github.com/NVIDIA/apex/blob/master/apex/optimizers/fused_lamb.py).

The following are what I did:

1.  For apex's lamb, use all default FusedLAMB settings (default: bias_correction=true, eps=1e-6, weight_decay=0.01).  

  For DeepSpeed's lamb, use the following:  bias_correction=true, eps=1e-6, weight_decay=0.01, max_coeff=10.0, min_coeff=0.01.

  For the final SQUAD finetune results, apex's results is about 8.0 higher than DeepSpeed's.

  From the loss curve, apex has much better converge than DeepSpeed's.
![e2e](https://user-images.githubusercontent.com/8562078/78373495-87ed6580-757f-11ea-8137-b8309f98bf11.png)

2.  By refering to [DeepSpeedSample bing bert code](https://github.com/microsoft/DeepSpeedExamples/blob/master/bing_bert/deepspeed_bsz16K_lamb_config.json), I changed DeepSpeed's lamb setting to
   bias_correction=false, eps=default, weight_decay=0.01, max_coeff=0.5, min_coeff=0.08.

 The loss cover looks better but still much worse than DeepSpeed's.
![bing](https://user-images.githubusercontent.com/8562078/78373537-95a2eb00-757f-11ea-9953-362d64c4aab9.png)

3. I tuned different values for bias_correction, max_coeff, min_coeff, learning rate, but couldn't make the converge better.

 So my question are: 
1) what are the recommended values to use DeepSpeed's FusedLamb? 
2) Can you provide something that work out of box for BERT?

**Comment by samyam** [comment]
@LiweiPeng  Thank you for trying out DeepSpeed. To narrow down what might be causing the loss gap can you provide us your bash script, bert config file and deepspeed config files?

Also, a few things that stand out to me are the following, can you give us some more information on the questions below:

1. Seems like your training is still in the early phases. The last step in the plot you shared is at 1387 which is about 45 million samples for batch size 32K. We expect a convergence at about 400M-500M samples. Can you share the loss curve for the entire training? 

2. Are you doing mixed sequence training? If so is this the seq length 128 portion of the curve?

3. The config you are referring to is for batch size 16K and for the seq 128 part of the training. After running seq 128, the training should be continued with seq length 512. At the end of the training (about 450M) samples, the F1 score for squad should be over 90.5. What is the score that you are seeing?

**Comment by LiweiPeng** [comment]
Thanks @samyam for the quick response. The following are more details.

1.  I am using the 2-phase training (max-seq-len=128 + max-seq-len=512). For the first loss curve, I finished the planned seq=128 part (total 14400 steps). Because the loss is so high and the SQUAD is so low, I didn't continue training for seq=512.   For the second loss-curve, I trained about 700 steps, then compared the loss curve. Because the loss didn't go down from 300 steps to 700 steps, I killed the training. I am not sure I could get a better converge if training to finish.

2. Yes, I am doing mixed seq training (2-phase training). They are both seq-len 128.

3. I didn't finish the training so didn't have final scores. As the example code is for batch size 16K, what settings need to be changed and how to change for batch size 32k?

I am using standard bert base config file downloaded from original google tf mode. The deepspeed config file for my second loss curve is like below
```
{
    "fp16":{
        "enabled":true,
        "loss_scale":0.0
    },
    "gradient_accumulation_steps":8,
    "gradient_clipping":1.0,
    "optimizer":{
        "params":{
            "bias_correction":false,
            "lr":0.00566,
            "max_coeff":0.5,
            "min_coeff":0.08,
            "weight_decay":0.01
        },
        "type":"lamb"
    },
    "prescale_gradients":true,
    "steps_per_print":1000000,
    "train_batch_size":null,
    "train_micro_batch_size_per_gpu":130,
    "wall_clock_breakdown":false,
    "zero_optimization":false
}
```

**Comment by samyam** [comment]
@LiweiPeng thanks for sharing the details. We are actively working on releasing a Bert pre-training tutorial. In the meanwhile, I can share the training schedule with you to try out. The schedule is for batch size 16K, so I would recommend using that batch size for now. The convergence is highly dependent on the schedule, so getting it right is super important. Please stay tuned for the full tutorial.

![image](https://user-images.githubusercontent.com/3409344/78411581-f4d71e80-75c4-11ea-8713-a149814c269a.png)

**Comment by samyam** [comment]
@LiweiPeng  also please note that we are not training seq 128 till 187000 steps, but only for 24430. The learning rate schedule is dependent on the max training steps value, so its important to set the max training steps correctly even if we are not training for fill 187K steps.

**Comment by LiweiPeng** [comment]
Thanks @samyam for the detailed parameters. This is very helpful. I'll give it a try.

**Comment by LiweiPeng** [comment]
Using the above bert parameters, I got expected results. I am closing this issue.
